// solidity-docgen/dist/handlebars.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.slug = exports.compile = exports.registerHelpers = void 0;
const handlebars_1 = __importDefault(require("handlebars"));
const H = handlebars_1.default.create();
H.registerHelper('slug', slug);
H.registerHelper({
    eq: function (v1, v2) {
        return v1 === v2;
    },
    ne: function (v1, v2) {
        return v1 !== v2;
    },
    lt: function (v1, v2) {
        return v1 < v2;
    },
    gt: function (v1, v2) {
        return v1 > v2;
    },
    lte: function (v1, v2) {
        return v1 <= v2;
    },
    gte: function (v1, v2) {
        return v1 >= v2;
    },
    and: function () {
        return [...arguments].slice(0, -1).every(Boolean);
    },
    or: function () {
        return [...arguments].slice(0, -1).some(Boolean);
    }
});
function registerHelpers(helpers) {
    H.registerHelper(helpers);
}
exports.registerHelpers = registerHelpers;
;
function compile(template) {
    const compiledTemplate = H.compile(template, { noEscape: true });
    return context => compiledTemplate(context, {
        allowProtoPropertiesByDefault: true,
        allowProtoMethodsByDefault: true,
    });
}
exports.compile = compile;
function slug(str) {
    return str.replace(/\W/g, '-');
}
exports.slug = slug;
